x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/evaluator/predicate/operator/comparison/EsqlBinaryComparison.java

```diff
@@ -32,14 +36,65 @@ public abstract class EsqlBinaryComparison extends BinaryComparison implements E
 
     private final Map<DataType, EsqlArithmeticOperation.BinaryEvaluator> evaluatorMap;
 
+    private final BinaryComparisonOperation functionType;
+
+    @FunctionalInterface
+    public interface BinaryOperatorConstructor {
+        EsqlBinaryComparison apply(Source source, Expression lhs, Expression rhs);
+    }
+
+    public enum BinaryComparisonOperation implements Writeable {
+
+        EQ(0, "==", BinaryComparisonProcessor.BinaryComparisonOperation.EQ, Equals::new),
+        // id 1 reserved for NullEquals
+        NEQ(2, "!=", BinaryComparisonProcessor.BinaryComparisonOperation.NEQ, NotEquals::new),
+        GT(3, ">", BinaryComparisonProcessor.BinaryComparisonOperation.GT, GreaterThan::new),
+        GTE(4, ">=", BinaryComparisonProcessor.BinaryComparisonOperation.GTE, GreaterThanOrEqual::new),
+        LT(5, "<", BinaryComparisonProcessor.BinaryComparisonOperation.LT, LessThan::new),
+        LTE(6, "<=", BinaryComparisonProcessor.BinaryComparisonOperation.LTE, LessThanOrEqual::new);
+
+        private final int id;
+        private final String symbol;
+        private final BinaryComparisonProcessor.BinaryComparisonOperation shim;
+        private final BinaryOperatorConstructor constructor;
+
+        BinaryComparisonOperation(
+            int id,
+            String symbol,
+            BinaryComparisonProcessor.BinaryComparisonOperation shim,
+            BinaryOperatorConstructor constructor
+        ) {
+            this.id = id;
+            this.symbol = symbol;
+            this.shim = shim;
+            this.constructor = constructor;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeVInt(id);
+        }
+
+        public static BinaryComparisonOperation readFromStream(StreamInput in) throws IOException {
+            int id = in.readVInt();
+            for (BinaryComparisonOperation op : values()) {
+                if (op.id == id) {
+                    return op;
+                }
+            }
+            throw new IOException("No BinaryComparisonOperation found for id [" + id + "]");
+        }
```

Comment:
This is a more normal way to serialize enums and let's us implement wire compatibility more easily. It's annoying to type it out for every single `enum`, but through hard won experience, we found it to be better.

Reply:
What is the advantage of having individual almost identical `readFromStream`/`writeTo` methods for every enum, as opposed to the current, centralized (in StreamInput/StreamOuput) approach?

I see the serialization tests make use of the transport version  (maybe this already exists now  in "old" serialization code, not sure) as a novelty, but can't we do the same in ESQL (outside StreamOutput/StreamInput) and, also, have a common Enum serialization/deserialization code?

Reply:
I can explain. We used to serialize enums the way that QL does - via the ordinals. It means fairly simple things like changing the order of the enums can break serialization. Or inserting a new enum. And adding wire backwards compatibility because a problem. You see here that we have to skip a number for wire compatibility. That's easy enough with the `id` member but requires an extra unused member. We broke enough enum serialization without realizing it that we decided we should be pedantic about it and make a method like this on each enum.

So, partly it's to remove the concerns around sorting enums or inserting new entries. And partly its to make wire backwards compatibility easier in the future. Well, in this case, now. And, partly, it's to make QLs enums look like the other well behaved serializable enums in Elasticsearch.

FWIW, I tend to use a `switch` statement instead of the loop. But that's not a big deal either way. I'd have done it like this:
```
        @Override
        public void writeTo(StreamOutput out) throws IOException {
            out.writeByte(code);
        }

        public static Role readFrom(StreamInput in) throws IOException {
            return switch (in.readByte()) {
                case 0 -> DEFAULT;
                case 1 -> INDEX_ONLY;
                case 2 -> SEARCH_ONLY;
                default -> throw new IllegalStateException("unknown role");
            };
        }
```

But it doesn't matter.

FWIW there are about 200 calls to `StreamOutput#writeEnum` and `StreamOutput#writeOptionalEnum` which does the ordinal thing. Spot checking enums in core yields a bunch that delegate to that and a bunch that do the `switch` or loop way. I think its not as settled of a thing as I make it out to be. But it really makes me feel good knowing I don't have to worry about changing the order of the constants breaking things over the wire.

Reply:
Thanks @nik9000

Reply:
I opened another issue to talk about removing the enum serialization this way. There's some discussion around doing some other third thing instead. I dunno. I was more sure about this a few hours ago. I mean, I'm sure it's an improvement over what we have, but I'm not sure there aren't other good ways too. OTOH, better is better.